closes #117

mpadge opened the pull request:
@maelle Here's an example from a package with a huge number of dependencies. Let me know what you think!

@noamross This adds an additional "*Package Dependencies*" section to the output - opinions very welcome, please!

---

## Checks for [lme4 (v1.1-27)](https://github.com/lme4/lme4/)

git hash: [6a6d121c](https://github.com/lme4/lme4//tree/6a6d121cd4fbf74da8e36b8c38a02a47724f0fdc)

- :heavy_check_mark: Package is already on CRAN.
- :heavy_multiplication_x: does not have a 'codemeta.json' file.
- :heavy_multiplication_x: does not have a 'contributing' file.
- :heavy_multiplication_x: does not use 'roxygen2'.
- :heavy_check_mark: 'DESCRIPTION' has a URL field.
- :heavy_check_mark: 'DESCRIPTION' has a BugReports field.
- :heavy_check_mark: Package has at least one HTML vignette
- :heavy_multiplication_x: These functions do not have examples: [checkConv, devcomp, glmerLaplaceHandle, glmFamily, isSingular, lme4_testlevel, lme4-package, lmResp, merPredD, mkMerMod, mkRespMod, mkSimulateTemplate, mkVarCorr, nlformula, pvalues, refitML, rePos, residuals.merMod, simulate.formula, troubleshooting].
- :heavy_multiplication_x:  Package uses global assignment operator ('<<-').
- :heavy_check_mark: Package has continuous integration checks.
- :heavy_multiplication_x: Package coverage failed
- :heavy_multiplication_x: R CMD check process failed with message: 'object 'cmdargs' not found'.

**Important:** All failing checks above must be addressed prior to proceeding

Package License: GPL (>=2)

---


### 1. Package Dependencies

<details>
<summary>Details of Package Dependency Usage (click to open)</summary>
<p>

The table below tallies all function calls to all packages ('ncalls'), both internal (r-base + recommended, along with the pacakge itself), and external (imported and suggested packages). 'NA' values indicate packages to which no identified calls to R functions could be found. Note that these results are generated by an automated code-tagging system which may not be entirely accurate.

|type       |package    | ncalls|
|:----------|:----------|------:|
|internal   |base       |   2613|
|internal   |lme4       |   1073|
|internal   |stats      |    323|
|internal   |methods    |     95|
|internal   |Matri      |      2|
|depends    |Matrix     |     12|
|imports    |utils      |     42|
|imports    |nlme       |     27|
|imports    |lattice    |     17|
|imports    |graphics   |     14|
|imports    |parallel   |      6|
|imports    |splines    |      5|
|imports    |MASS       |      5|
|imports    |boot       |      3|
|imports    |minqa      |      2|
|imports    |grid       |      1|
|imports    |nloptr     |      1|
|imports    |statmod    |      1|
|suggests   |ggplot2    |      2|
|suggests   |knitr      |     NA|
|suggests   |rmarkdown  |     NA|
|suggests   |PKPDmodels |     NA|
|suggests   |MEMSS      |     NA|
|suggests   |testthat   |     NA|
|suggests   |mlmRev     |     NA|
|suggests   |optimx     |     NA|
|suggests   |gamm4      |     NA|
|suggests   |pbkrtest   |     NA|
|suggests   |HSAUR3     |     NA|
|suggests   |numDeriv   |     NA|
|suggests   |car        |     NA|
|suggests   |dfoptim    |     NA|
|suggests   |mgcv       |     NA|
|linking_to |Rcpp       |     NA|
|linking_to |RcppEigen  |     NA|

---


</p></details>

### 2. Statistical Properties

This package features some noteworthy statistical properties which may need to be clarified by a handling editor prior to progressing.

<details>
<summary>Details of statistical properties (click to open)</summary>
<p>

The package has:

- code in C++ (23% in 6 files) and R (77% in 25 files)
- 4 authors
- 1  vignette
- 11 internal data files
- 12 imported packages
- 154 exported functions (median 13 lines of code)
- 606 non-exported functions in R (median 8 lines of code)
- 254 R functions (median 5 lines of code)

---

Statistical properties of package structure as distributional percentiles in relation to all current CRAN packages
The following terminology is used:

- `loc` = "Lines of Code"
- `fn` = "function"
- `exp`/`not_exp` = exported / not exported

The final measure (`fn_call_network_size`) is the total number of calls between functions (in R), or more abstract relationships between code objects in other languages. Values are flagged as "noteworthy" when they lie in the upper or lower 5th percentile.

|measure                  |  value| percentile|noteworthy |
|:------------------------|------:|----------:|:----------|
|files_R                  |     25|       86.3|           |
|files_src                |      6|       90.2|           |
|files_vignettes          |      7|       98.5|           |
|files_tests              |     82|       99.7|           |
|loc_R                    |   7628|       97.7|TRUE       |
|loc_src                  |   2244|       77.8|           |
|loc_vignettes            |   3593|       99.1|TRUE       |
|loc_tests                |   5394|       98.4|TRUE       |
|num_vignettes            |      1|       64.8|           |
|data_size_total          | 153096|       85.6|           |
|data_size_median         |    733|       62.4|           |
|n_fns_r                  |    760|       98.8|TRUE       |
|n_fns_r_exported         |    154|       97.7|TRUE       |
|n_fns_r_not_exported     |    606|       98.9|TRUE       |
|n_fns_src                |    254|       91.1|           |
|n_fns_per_file_r         |     16|       93.5|           |
|n_fns_per_file_src       |     30|       97.0|TRUE       |
|num_params_per_fn        |      3|       33.6|           |
|loc_per_fn_r             |      9|       24.3|           |
|loc_per_fn_r_exp         |     13|       30.5|           |
|loc_per_fn_r_not_exp     |      8|       22.6|           |
|loc_per_fn_src           |      5|        5.0|TRUE       |
|rel_whitespace_R         |      9|       93.1|           |
|rel_whitespace_src       |     16|       75.3|           |
|rel_whitespace_vignettes |      7|       88.3|           |
|rel_whitespace_tests     |     18|       98.2|TRUE       |
|doclines_per_fn_exp      |     92|       89.0|           |
|doclines_per_fn_not_exp  |      0|        0.0|TRUE       |
|fn_call_network_size     |   1395|       98.6|TRUE       |

---

</p></details>


### 2a. Network visualisation

Click to see the [interactive network visualisation of calls between objects in package](~/.cache/pkgcheck/static/lme4_pkgstats6a6d121c.html)

---

### 3. `goodpractice` and other checks

<details>
<summary>Details of goodpractice and other checks (click to open)</summary>
<p>

#### 3a. Continuous Integration Badges

[![R-CMD-check](https://github.com/lme4/lme4/workflows/R-CMD-check/badge.svg)](https://github.com/lme4/lme4/actions)

**GitHub Workflow Results**

|name                               |conclusion |sha    |date       |
|:----------------------------------|:----------|:------|:----------|
|.github/workflows/R-CMD-check.yaml |failure    |7c60db |2022-02-03 |
|R-CMD-check                        |skipped    |0bee8f |2022-03-19 |

---


#### 3b. `goodpractice` results


#### `R CMD check` with [rcmdcheck](https://r-lib.github.io/rcmdcheck/)

R CMD check generated the following error:

1. Error in stopifnot(no("func") || is.function(func), no("func") || is.list(args),  : 
  object 'cmdargs' not found


R CMD check generated the following check_fail:

1. no_description_depends


#### Test coverage with [covr](https://covr.r-lib.org/)

ERROR: Test Coverage Failed

#### Cyclocomplexity with [cyclocomp](https://github.com/MangoTheCat/cyclocomp)

Error : Build failed, unknown error, standard output:
[33m* checking for file ‘lme4/DESCRIPTION’ ... OK
* preparing ‘lme4’:
* checking DESCRIPTION meta-information ... OK
* cleaning src
* installing the package to process help pages
      -----------------------------------
ERROR: dependencies ‘minqa’, ‘nloptr’, ‘statmod’, ‘RcppEigen’ are not available for package ‘lme4’
* removing ‘/tmp/RtmpqbvYXS/Rinst140584d7deafd/lme4’
      -----------------------------------
ERROR: package installation failed
[39m


#### Static code analyses with [lintr](https://github.com/jimhester/lintr)

[lintr](https://github.com/jimhester/lintr) found the following 1222 potential issues:

message | number of times
--- | ---
Avoid 1:length(...) expressions, use seq_len. | 2
Avoid 1:ncol(...) expressions, use seq_len. | 1
Avoid 1:nrow(...) expressions, use seq_len. | 4
Avoid attach/detach, it is easy to create errors with it | 8
Avoid length(...):1 expressions, use seq_len. | 1
Avoid library() and require() calls in packages | 6
Avoid trailing semicolons, they are not needed. | 3
Avoid using sapply, consider vapply instead, that's type safe | 64
Lines should not be more than 80 characters. | 1127
Use <-, not =, for assignment. | 6



</p>
</details>

---

<details>
<summary>Package Versions</summary>
<p>

|package  |version   |
|:--------|:---------|
|pkgstats |0.0.3.95  |
|pkgcheck |0.0.2.275 |

</p>
</details>

mpadge commented:
Note that if and when this is integrated, we will have to address the long-standing task of updating the Dev Guide with explicit instructions for reviewers to pay attention to `pkgcheck` output.

maelle commented:
> Note that if and when this is integrated, we will have to address the long-standing task of updating the Dev Guide with explicit instructions for reviewers to pay attention to pkgcheck output.

Definitely!

I like the table, thank you! I have two suggestions but have no idea how feasible they are
- Could there be a list of functions?
- Could there be a list of places where the package is used? E.g. "R/, vignettes/articles/, others (inst/raw-data)".

noamross commented:
What's going on with the `Matri` package, which doesn't exist? Is that the test package?

Also, these are only _direct_ calls, correct? It does not go through the dependency tree?  (I don't think it necessarily should.  Dependency "weight" may be something to consider giving info about.)

mpadge commented:
Thanks both. @noamross:

> What's going on with the `Matri` package, which doesn't exist? Is that the test package?

Thanks for noting that - fixed in ropensci-review-tools/pkgstats#39

> Also, these are only _direct_ calls, correct? It does not go through the dependency tree? (I don't think it necessarily should. Dependency "weight" may be something to consider giving info about.)

Yep, definitely only direct calls.

---

@maelle:

> Could there be a list of functions?

That's a good idea, and would be quite useful to provide direct info on what a package actually does with all of its deps. I'm not sure the best way to do it. The `pkgstats` manuscript shows that most packages use only 1-2 functions from dependency packages, but some may use lots more. Plus i guess that info would be most useful when function usage from base + recommended packages was also included. The distribution of numbers of fns per pacakge will then be very strongly (Poisson/exponential) decreasing, so these can't be presented in any regular tabular form. They could all be one single "tidy" table, but that would generally be swamped by base-r, as the above example shows. My suggestion for the best solution would be have a bunch of additional `<details>` fold-outs, one for each package, with just a text-dump of function names and numbers of calls in brakcets. Shall I give that a try?

> Could there be a list of places where the package is used? E.g. "R/, vignettes/articles/, others (inst/raw-data)".

That information is all in the direct output from `pkgstats`, but difficult to present in any meaningful way in the `pkgcheck` report. My preference here - and indeed my aim - is to simply tell all developers/reviewers/editors that that information is readily accessible by just running `pkgstats()` locally - and that function does not require the package to be installed, so will always be painless enough to run.

maelle commented:
> My suggestion for the best solution would be have a bunch of additional <details> fold-outs, one for each package, with just a text-dump of function names and numbers of calls in brackets. Shall I give that a try?

Yes that sounds good!

> simply tell all developers/reviewers/editors that that information is readily accessible by just running pkgstats() locally - and that function does not require the package to be installed, so will always be painless enough to run.

In that case couldn't the names of functions used by package also be accessible via a function to be run locally?

mpadge commented:
@maelle After enormous amounts of debugging the html snapshots, the results of this PR now look like the following - let me know what you think of the function call details. Happy to modify any way you see fit. Thanks!

---


## Checks for [lme4 (v1.1-27)](https://github.com/lme4/lme4/)

git hash: [6a6d121c](https://github.com/lme4/lme4//tree/6a6d121cd4fbf74da8e36b8c38a02a47724f0fdc)

- :heavy_check_mark: Package is already on CRAN.
- :heavy_multiplication_x: does not have a 'codemeta.json' file.
- :heavy_multiplication_x: does not have a 'contributing' file.
- :heavy_multiplication_x: does not use 'roxygen2'.
- :heavy_check_mark: 'DESCRIPTION' has a URL field.
- :heavy_check_mark: 'DESCRIPTION' has a BugReports field.
- :heavy_check_mark: Package has at least one HTML vignette
- :heavy_multiplication_x: These functions do not have examples: [checkConv, devcomp, glmerLaplaceHandle, glmFamily, isSingular, lme4_testlevel, lme4-package, lmResp, merPredD, mkMerMod, mkRespMod, mkSimulateTemplate, mkVarCorr, nlformula, pvalues, refitML, rePos, residuals.merMod, simulate.formula, troubleshooting].
- :heavy_multiplication_x:  Package uses global assignment operator ('<<-').
- :heavy_check_mark: Package has continuous integration checks.
- :heavy_multiplication_x: Package coverage failed
- :heavy_multiplication_x: R CMD check process failed with message: 'object 'cmdargs' not found'.

**Important:** All failing checks above must be addressed prior to proceeding

Package License: GPL (>=2)

---


### 1. Package Dependencies

<details>
<summary>Details of Package Dependency Usage (click to open)</summary>
<p>

The table below tallies all function calls to all packages ('ncalls'), both internal (r-base + recommended, along with the package itself), and external (imported and suggested packages). 'NA' values indicate packages to which no identified calls to R functions could be found. Note that these results are generated by an automated code-tagging system which may not be entirely accurate.

|type       |package    | ncalls|
|:----------|:----------|------:|
|internal   |base       |   2679|
|internal   |lme4       |   1035|
|internal   |stats      |    341|
|internal   |methods    |    105|
|depends    |Matrix     |     13|
|imports    |utils      |     48|
|imports    |nlme       |     27|
|imports    |lattice    |     17|
|imports    |graphics   |     16|
|imports    |parallel   |      6|
|imports    |splines    |      5|
|imports    |MASS       |      5|
|imports    |boot       |      3|
|imports    |minqa      |      2|
|imports    |grid       |      1|
|imports    |nloptr     |      1|
|imports    |statmod    |      1|
|suggests   |ggplot2    |      2|
|suggests   |knitr      |     NA|
|suggests   |rmarkdown  |     NA|
|suggests   |PKPDmodels |     NA|
|suggests   |MEMSS      |     NA|
|suggests   |testthat   |     NA|
|suggests   |mlmRev     |     NA|
|suggests   |optimx     |     NA|
|suggests   |gamm4      |     NA|
|suggests   |pbkrtest   |     NA|
|suggests   |HSAUR3     |     NA|
|suggests   |numDeriv   |     NA|
|suggests   |car        |     NA|
|suggests   |dfoptim    |     NA|
|suggests   |mgcv       |     NA|
|linking_to |Rcpp       |     NA|
|linking_to |RcppEigen  |     NA|

Click below for tallies of functions used in each package. Locations of each call within this package may be generated locally by running 's <- pkgstats::pkgstats(<path/to/repo>)', and examining the 'external_calls' table.

<details>

<summary>base</summary>

<p>
list (410), if (206), c (194), lapply (97), is.null (92), attr (75), nrow (70), for (67), substitute (56), vapply (52), do.call (43), eval (43), seq_along (43), data.frame (38), t (29), drop (26), ncol (25), numeric (25), levels (24), sqrt (24), as.numeric (23), match.call (23), seq_len (23), diag (21), which (21), stopifnot (19), any (18), matrix (18), parent.frame (18), unlist (18), environment (17), list2env (17), sapply (17), colnames (16), rep.int (16), tryCatch (16), unname (16), call (15), is.na (15), rep (15), class (14), rbind (14), structure (13), inherits (12), sum (12), as.integer (11), deparse (11), dim (11), missing (11), paste (11), seq (11), setdiff (11), unique (11), UseMethod (11), all (10), beta (10), cbind (10), formals (10), max (10), abs (9), apply (9), as.data.frame (9), as.list (9), grep (9), array (8), diff (8), factor (8), get (8), is.numeric (8), log (8), norm (8), all.vars (7), as.name (7), getOption (7), match (7), nlevels (7), order (7), range (7), rownames (7), split (7), as.logical (6), as.vector (6), dimnames (6), gl (6), solve (6), chol (5), cumsum (5), lower.tri (5), new.env (5), scale (5), as.matrix (4), bquote (4), expand.grid (4), format (4), ifelse (4), is.character (4), labels (4), local (4), logical (4), Map (4), mapply (4), nchar (4), outer (4), row.names (4), sprintf (4), subset (4), switch (4), vector (4), warning (4), with (4), character (3), col (3), eval.parent (3), gamma (3), integer (3), is.matrix (3), mode (3), nzchar (3), quote (3), rev (3), round (3), substring (3), summary (3), Sys.getenv (3), tapply (3), tcrossprod (3), warnings (3), as.call (2), as.double (2), as.environment (2), crossprod (2), debug (2), eigen (2), emptyenv (2), exp (2), expression (2), getNamespaceExports (2), getRversion (2), gettextf (2), identical (2), identity (2), intersect (2), is.finite (2), is.list (2), ordered (2), parent.env (2), pmax (2), pmin (2), qr (2), Reduce (2), replicate (2), return (2), row (2), rowSums (2), stderr (2), sub (2), suppressMessages (2), suppressWarnings (2), sys.frame (2), system.time (2), table (2), union (2), withCallingHandlers (2), abbreviate (1), alist (1), as.expression (1), asNamespace (1), assign (1), attributes (1), by (1), cat (1), conditionCall (1), conditionMessage (1), force (1), grepl (1), gsub (1), I (1), is.data.frame (1), is.factor (1), is.logical (1), match.arg (1), mean (1), min (1), ngettext (1), parse (1), pi (1), pmatch (1), print (1), qr.coef (1), qr.X (1), Recall (1), sign (1), slice.index (1), sort (1), standardGeneric (1), stop (1), strsplit (1), svd (1), sys.frames (1), sys.function (1), system (1), try (1), upper.tri (1), which.max (1), zapsmall (1)
</p></details>
<details>

<summary>lme4</summary>

<p>
length (172), names (77), getME (34), sigma (29), zeta (27), ff (24), devfun (23), f (20), ss (16), paste0 (14), devcomp (12), findbars (11), lengths (11), isLMM (9), condVar (7), isGLMM (7), optwrap (7), RHSForm (7), deparse1 (6), dropOffset (6), gradient (6), mkGlmerDevfun (6), mkReTrms (6), nobars_ (6), spos (6), afun (5), factory (5), getConv (5), methTitle (5), mkdevfun (5), mlist2vec (5), ngrps (5), predy (5), reexpr (5), refit (5), vec2mlist (5), vv (5), checkCtrlLevels (4), factorize (4), fillmat (4), get.orig.levs (4), GHrule (4), grpfact (4), panel.thpr (4), subbars (4), wmsg (4), anyNA (3), arrange.condVar (3), asDf0 (3), barnames (3), checkConv (3), checkFormulaData (3), checkNlevels (3), checkScaleX (3), checkZdims (3), checkZrank (3), chkRank.drop.cols (3), Cv_to_Sv (3), devCrit (3), fe.zeta (3), forceCopy (3), gaussian_simfun (3), isNLMM (3), isREML (3), levelfun (3), llikAIC (3), lmer (3), mkLmerDevfun (3), mkNewReTrms (3), mkpar (3), mkPfun (3), mmList (3), namedList (3), nlformula (3), reOnly (3), abbrDeparse (2), allvarex (2), allVarsRec (2), bdiag_to_array (2), bdiag_to_mlist (2), binomial_simfun (2), bootMer (2), calc.vcov.hess (2), chck1 (2), checkHess (2), cont (2), cov2sdcor (2), Cv_to_Vv (2), dens (2), deriv12 (2), devfun2 (2), doCheck (2), expandDoubleVerts (2), famlink (2), fitfun (2), Gamma_simfun (2), get_re (2), getInds (2), getMsg (2), getNBdisp (2), getOptfun (2), getStart (2), glFormula (2), glmer (2), glmerPwrssUpdate (2), hasNoScale (2), hasScale (2), hessian (2), inverse.gaussian_simfun (2), lFormula (2), lme4_testlevel (2), logProf (2), ltsize (2), makeFac (2), makeInteraction (2), mkBlist (2), mkDataTemplate (2), mkParsTemplate (2), mkRespMod (2), mkVarCorr (2), modelFormula (2), mypred (2), NBfun (2), negative.binomial_simfun (2), Nelder_Mead (2), nextpar (2), nextstart (2), nlmerControl (2), nlminbwrap (2), nmkbw (2), nobars (2), nobs.merMod (2), optimizeGlmer (2), optTheta (2), panel.ci (2), poisson_simfun (2), pooledSD (2), prepanel.ci (2), print.merMod (2), profnames (2), refitML (2), refitNB (2), RglmerWrkIter (2), safe_chol (2), sdiag (2), Sv_to_Cv (2), tnames (2), Vv_to_Cv (2), ad (1), all0 (1), allFit (1), anovaLmer (1), anyBars (1), array_to_bdiag (1), as.data.frame.bootMer (1), as.data.frame.ranef.mer (1), as.data.frame.thpr (1), as.data.frame.VarCorr.merMod (1), as.function.merMod (1), attr_fun (1), augment.RE (1), bootFun (1), cat.f (1), cfun (1), check.boundary (1), checkArgs (1), checkResponse (1), chk.cconv (1), chk.convOpt (1), chooseFace (1), coef.lmList4 (1), coefMer (1), colSort (1), combineLists (1), confint.bootMer (1), confint.lmList4 (1), confint.merMod (1), confint.thpr (1), cooks.distance.influence.merMod (1), cooks.distance.merMod (1), deleteGroup (1), densityplot.thpr (1), deviance.merMod (1), df.residual.merMod (1), dfbeta.influence.merMod (1), dfbetas.influence.merMod (1), dim.merMod (1), dotplot.coef.mer (1), dotplot.ranef.mer (1), dp (1), drop1.merMod (1), dummy (1), est_theta (1), expandDoubleVert (1), expandSlash (1), extractAIC.merMod (1), family.glmResp (1), family.lmResp (1), family.merMod (1), family.nlsResp (1), family2char (1), fb (1), fitted.merMod (1), fixed.effects (1), fixef.merMod (1), format.perc (1), formatVC (1), formula.lmList4 (1), formula.merMod (1), frecurse (1), gamma.shape.merMod (1), get_clen (1), get.which (1), getCovariateFormula (1), getData.merMod (1), getFixedFormula (1), getGroups.lmList4 (1), getGroupsFormula (1), getGroupsFormula.merMod (1), getIDLabels (1), getME.merMod (1), getNBdisp.fam (1), glmer.nb (1), glmerLaplaceHandle (1), GQdk (1), grad.ctr (1), grad.ctr2 (1), grad.ctr3 (1), grad.ctr4 (1), hatvalues.merMod (1), influence.merMod (1), isAnyArgBar (1), isBar (1), isFlexLambda (1), isGLMlist (1), isGLMM.merMod (1), isLMM.merMod (1), isNBfamily (1), isNested (1), isNLMM.merMod (1), isRE (1), isREML.merMod (1), isSingular (1), lmList (1), log.thpr (1), logLik.merMod (1), lp (1), makeOp (1), matchNames (1), mcmcsamp (1), mcmcsamp.merMod (1), mkMerMod (1), mkMinimalData (1), mkOptCtrl (1), mlist_to_array (1), mmList.formula (1), mmList.merMod (1), model.frame.merMod (1), nafun (1), ngrps.default (1), ngrps.factor (1), ngrps.merMod (1), nlmer (1), noLHSform (1), normalizeFamilyName (1), npar.merMod (1), optimizeLmer (1), panel.blank (1), panelfun (1), plot.allFit (1), plot.coef.mer (1), plot.lmList4.confint (1), plot.ranef.mer (1), predict.merMod (1), print.allFit (1), print.bootMer (1), print.bootWarnings (1), print.ranef.mer (1), print.summary.merMod (1), print.VarCorr.merMod (1), profile.merMod (1), qqmath.ranef.mer (1), quickSimulate (1), random.effects (1), ranef.merMod (1), refit.merMod (1), refit2.merMod (1), refitML.merMod (1), reFormHack (1), REMLcrit (1), rep_len (1), rePCA (1), rePCA.merMod (1), residuals.glmResp (1), residuals.lmResp (1), sdcor2cov (1), simulate.formula_lhs_ (1), slashTerms (1), splitFormula (1), svals (1), tmpf (1), Tpfun (1), VarCorr.merMod (1), varex (1), vfun (1)
</p></details>
<details>

<summary>stats</summary>

<p>
formula (81), start (23), model.frame (21), logLik (16), fitted (14), predict (13), setNames (12), BIC (10), model.matrix (10), na.action (9), vcov (9), sd (7), var (7), gaussian (6), nobs (6), residuals (6), model.response (5), getCall (4), hat (4), hatvalues (4), na.omit (4), update (4), as.formula (3), coef (3), contrasts (3), df (3), model.offset (3), na.exclude (3), ave (2), binomial (2), coefficients (2), density (2), deviance (2), model.extract (2), model.weights (2), profile (2), qnorm (2), rbinom (2), resid (2), rnorm (2), step (2), asOneSidedFormula (1), confint (1), D (1), delete.response (1), df.residual (1), dfbeta (1), effects (1), end (1), Gamma (1), glm (1), inverse.gaussian (1), na.pass (1), nlminb (1), optim (1), optimize (1), pchisq (1), pf (1), poisson (1), qchisq (1), qt (1), quantile (1), simulate (1)
</p></details>
<details>

<summary>methods</summary>

<p>
el (20), is (15), initialize (12), new (12), setRefClass (10), setClass (9), as (7), extends (3), representation (3), setGeneric (3), setMethod (3), slotNames (3), rbind2 (2), getClass (1), getDataPart (1), slot (1)
</p></details>
<details>

<summary>utils</summary>

<p>
data (43), stack (2), head (1), methods (1), str (1)
</p></details>
<details>

<summary>nlme</summary>

<p>
fixef (17), ranef (6), VarCorr (4)
</p></details>
<details>

<summary>lattice</summary>

<p>
trellis.par.get (8), strip.custom (3), current.panel.limits (2), qqmath (2), llines (1), panel.number (1)
</p></details>
<details>

<summary>graphics</summary>

<p>
frame (13), abline (1), text (1), title (1)
</p></details>
<details>

<summary>Matrix</summary>

<p>
drop0 (4), rankMatrix (4), as.matrix (1), bdiag (1), Diagonal (1), fac2sparse (1), sparseMatrix (1)
</p></details>
<details>

<summary>parallel</summary>

<p>
clusterExport (2), mclapply (2), makeCluster (1), parLapply (1)
</p></details>
<details>

<summary>MASS</summary>

<p>
negative.binomial (5)
</p></details>
<details>

<summary>splines</summary>

<p>
interpSpline (4), backSpline (1)
</p></details>
<details>

<summary>boot</summary>

<p>
boot.ci (3)
</p></details>
<details>

<summary>ggplot2</summary>

<p>
facet_wrap (1), theme (1)
</p></details>
<details>

<summary>minqa</summary>

<p>
bobyqa (2)
</p></details>
<details>

<summary>grid</summary>

<p>
gpar (1)
</p></details>
<details>

<summary>nloptr</summary>

<p>
nloptr (1)
</p></details>
<details>

<summary>statmod</summary>

<p>
rinvgauss (1)
</p></details>

</p></details>

---


### 2. Statistical Properties

This package features some noteworthy statistical properties which may need to be clarified by a handling editor prior to progressing.

<details>
<summary>Details of statistical properties (click to open)</summary>
<p>

The package has:

- code in C++ (23% in 6 files) and R (77% in 25 files)
- 4 authors
- 1  vignette
- 11 internal data files
- 12 imported packages
- 154 exported functions (median 13 lines of code)
- 606 non-exported functions in R (median 8 lines of code)
- 254 R functions (median 5 lines of code)

---

Statistical properties of package structure as distributional percentiles in relation to all current CRAN packages
The following terminology is used:

- `loc` = "Lines of Code"
- `fn` = "function"
- `exp`/`not_exp` = exported / not exported

The final measure (`fn_call_network_size`) is the total number of calls between functions (in R), or more abstract relationships between code objects in other languages. Values are flagged as "noteworthy" when they lie in the upper or lower 5th percentile.

|measure                  |  value| percentile|noteworthy |
|:------------------------|------:|----------:|:----------|
|files_R                  |     25|       86.3|           |
|files_src                |      6|       90.2|           |
|files_vignettes          |      7|       98.5|           |
|files_tests              |     82|       99.7|           |
|loc_R                    |   7628|       97.7|TRUE       |
|loc_src                  |   2244|       77.8|           |
|loc_vignettes            |   3593|       99.1|TRUE       |
|loc_tests                |   5394|       98.4|TRUE       |
|num_vignettes            |      1|       64.8|           |
|data_size_total          | 153096|       85.6|           |
|data_size_median         |    733|       62.4|           |
|n_fns_r                  |    760|       98.8|TRUE       |
|n_fns_r_exported         |    154|       97.7|TRUE       |
|n_fns_r_not_exported     |    606|       98.9|TRUE       |
|n_fns_src                |    254|       91.1|           |
|n_fns_per_file_r         |     16|       93.5|           |
|n_fns_per_file_src       |     30|       97.0|TRUE       |
|num_params_per_fn        |      3|       33.6|           |
|loc_per_fn_r             |      9|       24.3|           |
|loc_per_fn_r_exp         |     13|       30.5|           |
|loc_per_fn_r_not_exp     |      8|       22.6|           |
|loc_per_fn_src           |      5|        5.0|TRUE       |
|rel_whitespace_R         |      9|       93.1|           |
|rel_whitespace_src       |     16|       75.3|           |
|rel_whitespace_vignettes |      7|       88.3|           |
|rel_whitespace_tests     |     18|       98.2|TRUE       |
|doclines_per_fn_exp      |     92|       89.0|           |
|doclines_per_fn_not_exp  |      0|        0.0|TRUE       |
|fn_call_network_size     |   1395|       98.6|TRUE       |

---

</p></details>


### 2a. Network visualisation

Click to see the [interactive network visualisation of calls between objects in package](/~/.cache/pkgcheck/static/lme4_pkgstats6a6d121c.html)

---

### 3. `goodpractice` and other checks

<details>
<summary>Details of goodpractice and other checks (click to open)</summary>
<p>

#### 3a. Continuous Integration Badges

[![R-CMD-check](https://github.com/lme4/lme4/workflows/R-CMD-check/badge.svg)](https://github.com/lme4/lme4/actions)

**GitHub Workflow Results**

|name                               |conclusion |sha    |date       |
|:----------------------------------|:----------|:------|:----------|
|.github/workflows/R-CMD-check.yaml |failure    |7c60db |2022-02-03 |
|R-CMD-check                        |skipped    |39e57d |2022-03-26 |

---


#### 3b. `goodpractice` results


#### `R CMD check` with [rcmdcheck](https://r-lib.github.io/rcmdcheck/)

R CMD check generated the following error:

1. Error in stopifnot(no("func") || is.function(func), no("func") || is.list(args),  : 
  object 'cmdargs' not found


R CMD check generated the following check_fail:

1. no_description_depends


#### Test coverage with [covr](https://covr.r-lib.org/)

ERROR: Test Coverage Failed

#### Cyclocomplexity with [cyclocomp](https://github.com/MangoTheCat/cyclocomp)

Error : Build failed, unknown error, standard output:
[33m* checking for file ‘lme4/DESCRIPTION’ ... OK
* preparing ‘lme4’:
* checking DESCRIPTION meta-information ... OK
* cleaning src
* installing the package to process help pages
      -----------------------------------
ERROR: dependencies ‘minqa’, ‘nloptr’, ‘statmod’, ‘RcppEigen’ are not available for package ‘lme4’
* removing ‘/tmp/RtmpqbvYXS/Rinst140584d7deafd/lme4’
      -----------------------------------
ERROR: package installation failed
[39m


#### Static code analyses with [lintr](https://github.com/jimhester/lintr)

[lintr](https://github.com/jimhester/lintr) found the following 1222 potential issues:

message | number of times
--- | ---
Avoid 1:length(...) expressions, use seq_len. | 2
Avoid 1:ncol(...) expressions, use seq_len. | 1
Avoid 1:nrow(...) expressions, use seq_len. | 4
Avoid attach/detach, it is easy to create errors with it | 8
Avoid length(...):1 expressions, use seq_len. | 1
Avoid library() and require() calls in packages | 6
Avoid trailing semicolons, they are not needed. | 3
Avoid using sapply, consider vapply instead, that's type safe | 64
Lines should not be more than 80 characters. | 1127
Use <-, not =, for assignment. | 6



</p>
</details>

---

<details>
<summary>Package Versions</summary>
<p>

|package  |version   |
|:--------|:---------|
|pkgstats |0.0.3.96  |
|pkgcheck |0.0.2.290 |

</p>
</details>

maelle commented on tests/testthat/_snaps/extra-checks/checks-extra.md:
```diff
@@ -21,7 +21,51 @@ Package License: GPL-3
+|suggests   |testthat   |     NA|
+|linking_to |cpp11      |     NA|
+
+Click below for tallies of functions used in each package. Locations of each call within this package may be generated locally by running 's <- pkgstats::pkgstats(<path/to/repo>)', and examining the 'external_calls' table.
```
what is "below"

maelle reviewed (approved):
Thank you!

mpadge commented on R/format-checks.R:
```diff
@@ -184,7 +191,56 @@ checks_to_markdown <- function (checks, render = FALSE) {
+        "<p>",
+        "",
+        paste0 (
+            "The table below tallies all function calls to all packages ",
```
.. I went to change that, but actually that makes the English gramatically incorrect, so i just can't bring myself to do it.

mpadge commented on R/pkg-deps.R:
```diff
@@ -0,0 +1,160 @@
+    op <- grep ("\\(", nmsp)
+    cl <- grep ("\\)", nmsp)
+    if (length (op) != length (cl)) {
+        stop ("'NAMESPACE' file is incorrectly formatted")
```
Nope, but enough NAMESPACE files are (still) hand-written that would be possible.

mpadge commented on tests/clean-snapshots.R:
```diff
@@ -28,6 +28,25 @@ edit_markdown <- function (md) {
+        md <- md [-index]
+    }
+
+    # and for some reason, the covr environment pulls in an extra external
```
yeah, `covr` is a nightmare for software which intro-spects packages, because it changes the underlying source code. And does so in a cloned environment of the whole package, so lots of package-environment expectations also break.

mpadge commented on R/format-checks.R:
```diff
@@ -483,16 +539,20 @@ render_markdown <- function (md, open = TRUE) {
 
-    fmd <- tempfile (pattern = "pkgcheck", fileext = ".Rmd")
+    f <- tempfile (pattern = "pkgcheck")
+    fmd <- paste0 (f, ".Rmd")
```
Thanks, done
